stl/inc/vector

```diff
@@ -578,9 +581,15 @@ private:
     }
 
 public:
-    vector(vector&& _Right) noexcept : _Mypair(_One_then_variadic_args_t{}, _STD move(_Right._Getal())) {
+    vector(vector&& _Right) noexcept
+        : _Mypair(_One_then_variadic_args_t{}, _STD move(_Right._Getal()), _Right._Mypair._Myval2._Myfirst,

```

Comment:
I assume we cannot use `_STD exchange(_Right._Mypair._Myval2._Myfirst, nullptr)` or doe we have some internal `exchange` version that is always available

Reply:
~~`exchange` is defined in `<utility>` which this header already includes (`<xmemory>` -> `<xutility>` -> `<utility>`) so we could use it here. I wouldn't reject the PR for not doing so, but I am a fan of the slightly higher level vocabulary of `meow = exchange(x, y)` over `meow = x; x = y;`.~~

Ouch: we can't easily null the pointers in the source until we know the allocation of the proxy on line 587 (the last potentially failing operation) has succeeded. Allocation failure would leak the element memory, and leave the source vector in a nonsense state.

There's a different design in which this constructor allocates the proxy and then delegates to another constructor that does the real initialization (including `exchange` calls), but it's not worth the trouble of that much refactoring so we can use `exchange`.

Reply:
>There's a different design in which this constructor allocates the proxy and then delegates to another constructor that does the real initialization

That has horrible codegen consequences unfortunately :(

Reply:
> That has horrible codegen consequences unfortunately :(

That bug will be turning 4 next month.

Reply:
@CaseyCarter Note that this ctor is mandated to be `noexcept`; our pre-existing behavior is that proxy allocation failure will terminate.

Reply:
> @CaseyCarter Note that this ctor is mandated to be `noexcept`; our pre-existing behavior is that proxy allocation failure will terminate.

True; I retract my prophecies of doom and am now once again mildly in favor of `exchange` here.

Reply:
@CaseyCarter and I looked into the codegen here - I was surprised to find that `exchange` actually improves the codegen (on x86), making it 2 bytes smaller. This is using `exchange` directly in the mem-initializer-list without a delegating constructor. I went ahead and pushed one more commit for this.

Reply:
@StephanTLavavej and I spent some time poring over x86 codegen diffs with and without `exchange` to ensure the change wouldn't damage anything. Surprisingly the codegen was _better_ with `exchange`: the compiler uses one fewer register temporaries saving a couple of bytes to push and pop the old register contents.

Reply:
Such a good story it was worth telling twice.